Task assemblies can mismatch between .NET versions

Ref: https://discord.com/channels/732297728826277939/732297808148824115/1205082594757054494

```

Filip Navara — Today at 10:27 AM

    88>C:\Program Files\dotnet\packs\Microsoft.iOS.Sdk\17.2.8004\tools\msbuild\iOS\Xamarin.Shared.targets(546,3): error MSB4018: The "CompileAppManifest" task failed unexpectedly.
    88>C:\Program Files\dotnet\packs\Microsoft.iOS.Sdk\17.2.8004\tools\msbuild\iOS\Xamarin.Shared.targets(546,3): error MSB4018: System.MissingMethodException: Method not found: 'System.String Xamarin.Localization.MSBuild.MSBStrings.get_W7108()'.
    88>C:\Program Files\dotnet\packs\Microsoft.iOS.Sdk\17.2.8004\tools\msbuild\iOS\Xamarin.Shared.targets(546,3): error MSB4018:    at Xamarin.MacDev.Tasks.CompileAppManifestTaskBase.Execute()
    88>C:\Program Files\dotnet\packs\Microsoft.iOS.Sdk\17.2.8004\tools\msbuild\iOS\Xamarin.Shared.targets(546,3): error MSB4018:    at Xamarin.MacDev.Tasks.CompileAppManifest.Execute() in /Users/builder/azdo/_work/1/s/xamarin-macios/msbuild/Xamarin.MacDev.Tasks/Tasks/CompileAppManifest.cs:line 16
    88>C:\Program Files\dotnet\packs\Microsoft.iOS.Sdk\17.2.8004\tools\msbuild\iOS\Xamarin.Shared.targets(546,3): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
    88>C:\Program Files\dotnet\packs\Microsoft.iOS.Sdk\17.2.8004\tools\msbuild\iOS\Xamarin.Shared.targets(546,3): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
    
    does this ring any bells? .NET 8.0.200-preview.23624.5 SDK, VS 17.9.0 Preview 4.0
    (I know it's a "missing file" error, not really asking about the underlying cause, just the localization bundling issue)

Rolf — Today at 10:31 AM

    that's weird, W7108 is in the initial .NET 8 release
    is this a .NET 7 project?
    or are there .NET 7 projects in the solution?

Filip Navara — Today at 10:32 AM

    It was .NET 7 project and I bumped it to .NET 8.
    There are other .NET 7 projects in the solution, possibly even some iOS library projects
    so yeah, it would make sense that it can collide within the same VS instance

Filip Navara — Today at 10:40 AM

    yep, goes away if I retarget everything and restart VS
    (in fact, even the build error goes away)
```

See also https://github.com/xamarin/xamarin-macios/issues/9835, where a fix was attempted (but apparently not enough).

--- a/msbuild/Messaging/Xamarin.Messaging.Build/Xamarin.Messaging.Build.csproj
+++ b/msbuild/Messaging/Xamarin.Messaging.Build/Xamarin.Messaging.Build.csproj
@@ -39,6 +39,9 @@
       <Reference Include="Microsoft.Build.Utilities.Core">
           <HintPath>$(MonoMSBuildBinPath)/Microsoft.Build.Utilities.Core.dll</HintPath>
       </Reference>
+      <Compile Include="..\..\Versions.g.cs">
+         <Link>Versions.g.cs</Link>
+      </Compile>
   </ItemGroup>
 
   <ItemGroup>

--- a/msbuild/Xamarin.Localization.MSBuild/Properties/AssemblyInfo.cs
+++ b/msbuild/Xamarin.Localization.MSBuild/Properties/AssemblyInfo.cs
@@ -1,0 +1,10 @@
+using System;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+[assembly: AssemblyCompanyAttribute ("Microsoft Corp.")]
+[assembly: AssemblyFileVersionAttribute (VersionConstants.XamarinIOSVersion)]
+[assembly: AssemblyInformationalVersionAttribute (VersionConstants.XamarinIOSVersion + "." + VersionConstants.NuGetPrereleaseIdentifier + "+" + VersionConstants.NuGetBuildMetadata)]
+[assembly: AssemblyProductAttribute ("Xamarin.Localization.MSBuild")]
+[assembly: AssemblyTitleAttribute ("Xamarin.Localization.MSBuild")]
+[assembly: AssemblyVersionAttribute (VersionConstants.XamarinIOSVersion)]

--- a/msbuild/Xamarin.Localization.MSBuild/Xamarin.Localization.MSBuild.csproj
+++ b/msbuild/Xamarin.Localization.MSBuild/Xamarin.Localization.MSBuild.csproj
@@ -7,6 +7,7 @@
     <Nullable>enable</Nullable>
     <LangVersion>latest</LangVersion>
     <WarningsAsErrors>Nullable</WarningsAsErrors>
+    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
   </PropertyGroup>
   <PropertyGroup>
     <EnableDefaultEmbeddedResourceItems>false</EnableDefaultEmbeddedResourceItems>
@@ -27,6 +28,9 @@
     </EmbeddedResource>
   </ItemGroup>
   <ItemGroup>
+    <Compile Include="..\Versions.g.cs">
+      <Link>Versions.g.cs</Link>
+    </Compile>
     <EmbeddedResource Include="TranslatedAssemblies/MSBStrings.cs.resx">
       <Link>MSBStrings.cs.resx</Link>
       <ManifestResourceName>Xamarin.Localization.MSBuild.MSBStrings.cs</ManifestResourceName>

--- a/msbuild/Xamarin.MacDev.Tasks/Xamarin.MacDev.Tasks.csproj
+++ b/msbuild/Xamarin.MacDev.Tasks/Xamarin.MacDev.Tasks.csproj
@@ -85,9 +85,6 @@
     <Compile Include="..\..\tools\common\NullableAttributes.cs">
       <Link>external\NullableAttributes.cs</Link>
     </Compile>
-    <Compile Include="..\Versions.g.cs">
-      <Link>Versions.g.cs</Link>
-    </Compile>
     <Compile Remove="Errors.designer.cs" /> <!-- The 'CoreResGen' target will add it again from the EmbeddedResource item, this avoids a warning about the file being compiled twice -->
     <Compile Include="..\..\tools\common\SdkVersions.cs">
       <Link>external\SdkVersions.cs</Link>

--- a/tools/devops/automation/templates/build/api-diff-process-results.yml
+++ b/tools/devops/automation/templates/build/api-diff-process-results.yml
@@ -69,7 +69,7 @@ steps:
       Import-Module $Env:SYSTEM_DEFAULTWORKINGDIRECTORY\xamarin-macios\tools\devops\automation\scripts\MaciosCI.psd1
       $vsdropsChangeDetectionPrefix = "https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/detected-changes/$Env:BUILD_BUILDNUMBER/$Env:BUILD_BUILDID-$Env:SYSTEM_JOBATTEMPT/;/"
 
-      $rootDirectory = Join-Path "$Env:BUILD_ARTIFACTSTAGINGDIRECTORY" "change-detection" "results"
+      $rootDirectory = Join-Path "$Env:SYSTEM_DEFAULTWORKINGDIRECTORY" "change-detection" "results"
 
       $inputContentsPath = Join-Path -Path $rootDirectory -ChildPath "gh-comment.md"
       if (Test-Path $inputContentsPath -PathType leaf) {

